docs/migration-guides/v2/lib.md

```diff
@@ -0,0 +1,412 @@
+
+**Note:** This section is only relevant if you’re using the `RealtimeChannel.whenState()` or `Connection.whenState()` methods.
+
+The `RealtimeChannel.whenState()` and `Connection.whenState()` methods now return `null` when the connection is already in the given state, instead of attempting to synthesize an artificial state change.
+
+#### Be aware that the `fromEncoded()` and `fromEncodedArray()` methods are now async
+
+**Note:** This section is only relevant if you’re using `Message` or `PresenceMessage`’s `fromEncoded` or `fromEncodedArray` methods.
+
+`Message` and `PresenceMessage`’s `fromEncoded` and `fromEncodedArray` methods now operate asynchronously. That is, instead of returning the decoding result, they return a promise. Update your code to retrieve the result of these promises, for example by using the `await` keyword.
+
+<h4 id="secure-context">Be aware that symmetric encryption in a browser now requires a secure context</h4>
+
+**Note:** This section is only relevant if you’re using the `cipher` client option and running in a browser.
+
+If you’re making use of the `cipher` client option to enable symmetric encryption on a channel, be aware that when running in a browser this functionality is now implemented using the [Web Crypto API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Crypto_API). Hence, this functionality is only available when this API is available; namely, when the current environment is a [secure context](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts). Roughly speaking, this means that you can now only use Ably channel encryption inside a web page when serving your page over HTTPS or from `localhost`.
+
+<h4 id="no-more-CryptoJS">Be aware that the library no longer uses CryptoJS and hence no longer works with <code>WordArray</code></h4>
+
+**Note:** This section is only relevant if you’re using the `cipher` client option and running in a browser.
+
+When running in a browser, the library previously used the CryptoJS library for implementing symmetric encryption on a channel. As mentioned [here](#secure-context), the library now instead uses the built-in Web Crypto API.
+
+This means that the library is no longer capable of interacting with CryptoJS’s `WordArray` type. Specifically:
+
+- `Crypto.generateRandomKey()` now returns an `ArrayBuffer` instead of a `WordArray`.
+- `Crypto.getDefaultParams({ key })` no longer accepts a `WordArray` key; pass an `ArrayBuffer` or `Uint8Array` instead.
+
+#### Stop requesting the `xhr_streaming` and `xhr` transports
+
+**Note:** This section is only relevant if you’re explicitly writing `"xhr_streaming"` or `"xhr"` as part of your `transports` client option.
+
+The `xhr_streaming` transport, also known as `xhr`, has been removed. If your `transports` client option contains `"xhr_streaming"` or `"xhr"`, remove this value. If you still wish to explicitly request a non-WebSocket transport, request `xhr_polling` instead.
```

Comment:
@owenpearson (probably actually @SimonWoolf since Owen is away now) do you have any further information on why `xhr_streaming` was removed? I’m telling users to use `xhr_polling` instead, but that’s worse (in terms of responsiveness), right? I’m just a bit out of the loop on this change

Reply:
We're now doing fallback instead of upgrade. Say you try to connect with websocket, it times out, you try the health check endpoints, which tells you that there's internet connectivity but websockets are broken for some reason. So that likely means you're being MITM'd by some shitty corporate proxy that doesn't deal with websockets properly, and the guarantees that we'd normally get with TLS are worthless. So now you have to ask yourself: what transport should you try next? Because we know from the whole e*****a saga that corporate proxy that break websockets can break xhr streaming as well! We need to try something that we know will work even with the most basic of shitty web proxies, which means long polling.

Given which, what role is left for xhr streaming?

Of course not using it in the default connection strategy doesn't necessarily mean removing it from the lib altogether. Not sure if there was an explicit discussion on that or if owen just figured that if we're not using it by default then we should just remove it. Shrug, if some customer comes to us and says "We're using a very particular corporate firewall that can't cope with websockets but actually can do xhr streaming fine" then we can always re-add it

more background: https://github.com/ably/ably-js/issues/1330

Reply:
> shitty corporate proxies that break websockets probably also break break xhr streaming

Ah, I didn’t know this!

> Of course not using it in the default connection strategy doesn't necessarily mean removing it from the lib altogether. Not sure if there was an explicit discussion on that or if owen just figured that if we're not using it by default then we should just remove it.

Yeah, that's the part that wasn't clear to me. Well, it's done now, and Owen’s not here to explain, and there don't seem to be strong opinions about it, so I guess we'll be sticking with it.

> Shrug, if some customer comes to us and says "We're using a very particular corporate firewall that can't cope with websockets but actually can do xhr streaming fine" then we can always re-add it

OK, cool.

So sounds like there isn't really else that I can add to the migration guide (explaining why we removed it, that is); let's see if anyone complains 🤷

Reply:
Actually, it sounds like, given that it didn't help massively with firewalls, the main reason that `xhr_streaming` existed was to provide a WebSocket-like experience in browsers that didn't support WebSocket. With that in mind, I've updated this guidance:

> v1 offered the `xhr_streaming` transport, also known as `xhr`, to provide a performant realtime transport for browsers which did not support WebSocket connections. Since this limitation does not apply to any of the [browsers supported by v2](#supported-platforms), the `xhr_streaming` transport has been removed. If your `transports` client option contains `"xhr_streaming"` or `"xhr"`, remove this value. If you still wish to explicitly request a non-WebSocket transport, request `xhr_polling` instead.